python/GafferUI/CompoundEditor.py

```diff
@@ -495,3 +514,309 @@ def __removePanel( self ) :
 		# issues in certain circumstances (eg hosting Gaffer in other Qt DCC).
 		## \todo: consider doing this for all Menu commands in GafferUI.Menu
 		GafferUI.EventLoop.addIdleCallback( functools.partial( Gaffer.WeakMethod( splitContainerParent.join ), 1 - splitContainerParent.index( splitContainer ) ) )
+
+
+
+## An internal eventFilter class managing all tab drag-drop events and logic.
+# Tab dragging is and exception and is implemented entirely using native Qt
+# events. This was decided after trying all combinations of Qt-only/Gaffer-only
+# and Gaffer/Qt hybrid as:
+#
+#  - We wanted the visual sophistication of QTabBar.setMovable() for
+#    re-arranging tabs within their parent container.
+#  - Qt doesn't use its own drag mechanism for setMoveable so we have to
+#    interact at a lower level that we'd otherwise like.
+#  - We can't programmatically start a Gaffer drag.
+#  - We had issues propagating style-changes during a Gaffer drag.
+#
+# As such, it is somewhat of an outlier compared to how the rest of the
+# application works. To avoid polluting the code base, it has been structured
+# so that as much logic and Qt interaction is contained within this class.
+#
+# The broad responsibilities of the _TabDragEventManager are:
+#
+#  - Capture the starting state of any TabBar drag.
+#  - Abort the built-in tab-move behaviour of QTabBar when the cursor leaves
+#    the logical are of the TabbedContainer in-which the tab lives.
+#  - Manage hover states during a drag operation.
+#  - Coordinate the re-parenting of Tab widgets.
+#  - Manage the life-cycle of detached panels affected by the drag.
+#
+class _TabDragEventsManager( QtCore.QObject ) :
+
+	__tabMimeType = "application/x-gaffer.compoundEditor.tab"
+
+	def __init__( self, tabbedContainer ) :
+
+		QtCore.QObject.__init__( self )
+
+		self.__eventMask = set ( (
+			QtCore.QEvent.MouseButtonPress,
+			QtCore.QEvent.MouseButtonRelease,
+			QtCore.QEvent.MouseMove,
+			QtCore.QEvent.DragEnter,
+			QtCore.QEvent.DragLeave,
+			QtCore.QEvent.Drop
+		) )
+
+		self.__abortedInitialRearrange = False
+
+		self.__tabbedContainer = weakref.ref( tabbedContainer )
```

Comment:
@johnhaddon As per our earlier discussion, this is the 'tidier' version, where, as we consider the _TabDragEventsManager an extended presence of the parent _TabbedContainer (rather than a free radical), holding a ref here is perhaps worth to simplify some of the state management of the tabbed container that the manager performs. Happy to do a commit where we remove the weakref and go find it where it's required if this is preferred.